### PR TITLE
Add eager-loading

### PR DESCRIPTION
**Pull request by stibiumz**
Some people have requested me to add [this package](/stibiumz/phalcon.eager-loading) to the Incubator, so here it's

### CODE REVIEW

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 1]
```diff
@@ -0,0 +1,243 @@
+<?php namespace Phalcon\Mvc\Model\EagerLoading;
+
+use Phalcon\Mvc\Model\Relation,
+	Phalcon\Mvc\Model\Resultset,
+	Phalcon\Mvc\Model\Query\Builder;
+
+/**
+ * Represents a level in the relations tree to be eagerly loaded
+ */
+final class EagerLoad {
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: Can you fix this? 
We use PSR-2 and tests too

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+	private $subject;
+	/** @var boolean */
+	static private $isPhalcon2;
+
+	/**
+	 * @param RelationInterface
+	 * @param null|callable $constraints
+	 * @param Loader|EagerLoad $parent
+	 */
+	public function __construct(Relation $relation, $constraints, $parent) {
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: and this

**Comment by sergeyklay** [comment]
Fine! Thank you for contributing. Can you provide a small test please?

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+		// PHQL has problems with this slash
+		if ($relReferencedModel[0] === '\\') {
+			$relReferencedModel = ltrim($relReferencedModel, '\\');
+		}
+
+		$bindValues = [];
+
+		foreach ($this->parent->getSubject() as $record) {
+			$bindValues[$record->readAttribute($relField)] = TRUE;
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: `TRUE` → `true`

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+					$record->{$alias} = $referencedModels;
+
+					if (static::$isPhalcon2) {
+						$record->{$alias} = NULL;
+						$record->{$alias} = $referencedModels;
+					}
+				}
+				else {
+					$record->{$alias} = NULL;
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: `NULL` → `null`

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+	/**
+	 * @param ModelInterface|ModelInterface[]|Simple $from
+	 * @param ...$arguments
+	 * @throws \InvalidArgumentException
+	 */
+	public function __construct($from) {
+		$error     = FALSE;
+		$arguments = array_slice(func_get_args(), 1);
+
+		if (! $from instanceof ModelInterface) {
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: `if (! $` → `if (!$`

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+			$this->mustReturnAModel = TRUE;
+		}
+
+		if ($error) {
+			throw new \InvalidArgumentException(static::E_INVALID_SUBJECT);
+		}
+
+		$this->subject          = $from;
+		$this->subjectClassName = $className;
+		$this->eagerLoads       = empty ($arguments) ? array () : static::parseArguments($arguments);
```
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: `empty (` → `empty(`

**Comment by stibiumz** [comment]
Yeah, no problem, I'll fix it.
About the tests, how I do it? A database it's needed, Do I need to create it with Travis CI? I'm pretty new with this

**Comment by sergeyklay** [comment]
@stibiumz `master` is freezed. Create PR to `2.0.x` please.

About DB tests http://docs.travis-ci.com/user/database-setup/#MySQL
Feel free to ask any question if any.

**Comment by stibiumz** [comment]
Ok, thanks. Today it's late for me, tomorrow I'll read the documentation and create a new PR :+1:

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 7]
```diff
@@ -0,0 +1,243 @@
+<?php namespace Phalcon\Mvc\Model\EagerLoading;
+
+use Phalcon\Mvc\Model\Relation,
+	Phalcon\Mvc\Model\Resultset,
+	Phalcon\Mvc\Model\Query\Builder;
```
Comment (Green-Cat) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: as QueryBuilder;

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+	 * @param null|callable $constraints
+	 * @param Loader|EagerLoad $parent
+	 */
+	public function __construct(Relation $relation, $constraints, $parent) {
+		if (static::$isPhalcon2 === NULL) {
+			static::$isPhalcon2 = version_compare(\Phalcon\Version::get(), '2.0.0') >= 0;
+		}
+
+		$this->relation    = $relation;
+		$this->constraints = is_callable($constraints) ? $constraints : NULL;
```
Comment (Green-Cat) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: NULL -> null

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+		$bindValues = [];
+
+		foreach ($this->parent->getSubject() as $record) {
+			$bindValues[$record->readAttribute($relField)] = TRUE;
+		}
+
+		$bindValues = array_keys($bindValues);
+
+		$subjectSize         = count($this->parent->getSubject());
+		$isManyToManyForMany = FALSE;
```
Comment (Green-Cat) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: FALSE -> false

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+							$relReferencedField
+						)
+					)
+					->inWhere("[{$relIrModel}].[{$relIrField}]", $bindValues)
+				;
+			}
+			else {
+				// The query is for many models, so it's needed to execute an
+				// extra query
+				$isManyToManyForMany = TRUE;
```
Comment (Green-Cat) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: TRUE -> true

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php [hunk 11]
```diff
@@ -0,0 +1,243 @@
+<?php namespace Phalcon\Mvc\Model\EagerLoading;
```
Comment (Green-Cat) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: New line

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php (hunk 5 above)
Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: This is not specified in PSR2

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php (hunk 6 above)
Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: This also is not specified in PSR2, `empty` is not a function, lt's a [language construct](http://php.net/manual/en/reserved.keywords.php)

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php (hunk 7 above)
Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: Already in use

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php (hunk 11 above)
Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: Not specified in PSR2

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php (hunk 5 above)
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: http://www.php-fig.org/psr/psr-2/

> Opening parentheses for control structures MUST NOT have a space after them, and closing parentheses for control structures MUST NOT have a space before.

Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: This is not the same, but usually all follow this rule

Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: And where's the space after the opening parenthesis?

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php (hunk 7 above)
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: Correct PSR CS:

```php
use Phalcon\Mvc\Model\Relation;
use Phalcon\Mvc\Model\Resultset;
use Phalcon\Mvc\Model\Query\Builder;
```

`use` per line

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php (hunk 11 above)
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: @stibiumz 
Ok. Let's test this:

```sh
$ cat test.php 
```

**Test 1:**

```php
<?php namespace Foo\Bar;
/**
 * Doc
 *
 * PHP version 5
 *
 * @category Bar
 * @package  Foo
 * @author   Display Name <username@example.com>
 * @license  MIT http://mit.com
 * @link     http://site.com
 */

if (!false) {
    
}
```

run test

```sh
$ phpcs test.php  # phpcs — Is a PHP Code Sniffer


FILE: /srv/http/test/test.php
----------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
----------------------------------------------------------------------
 1 | ERROR | Missing file doc comment
----------------------------------------------------------------------

Time: 23ms; Memory: 1.5Mb

```

**Test 2:**

```sh
$ cat test.php 
```

```php
<?php
/**
 * Doc
 *
 * PHP version 5
 *
 * @category Bar
 * @package  Foo
 * @author   Display Name <username@example.com>
 * @license  MIT http://mit.com
 * @link     http://site.com
 */

namespace Foo\Bar;

if (!false) {
    
}
```

run test

```sh
$ phpcs test.php  # phpcs — Is a PHP Code Sniffer
# ok
```

File: Library/Phalcon/Mvc/Model/EagerLoading/Loader.php (hunk 5 above)
Comment (sergeyklay) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: @stibiumz 
You're right. But:

> This is not the same, but usually all follow this rule

This is just common incubator style.

Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/Loader.php]: I answered late, I'll fix it then :+1:

File: Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php (hunk 11 above)
Comment (stibiumz) [inline Library/Phalcon/Mvc/Model/EagerLoading/EagerLoad.php]: Shouldn't it be `$ phpcs --standard=PSR2 test.php`?

If so:

**Pull request closed without merging**